cctools-llvm: use cctools assembler on LLVM 11 and x86_64-darwin

reckenrode opened the pull request:
###### Description of changes

This PR addresses the build failure in https://github.com/NixOS/nixpkgs/pull/241951#issuecomment-1625726447 more generally by using the cctools assembler with LLVM 11 on x86_64-darwin (in case there are other packages that are sensitive to the differences in output it can produce depending on debug information).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

reckenrode committed:
cctools-llvm: use cctools assembler on x86_64-darwin and LLVM 11

Clang 11 performs an optimization on x86_64 that is sensitive to the
presence of debug information. This results in GCC’s bootstrap failing
because it builds stage 2 with debug information and stage 3 without,
and the resulting objects do not match.

This patch uses the cctools assembler on LLVM 11 and x86_64-darwin while
using the integrated assembler on newer versions, which matches the
platform tools (Apple has uses the integrated assembler since Xcode 12).

reckenrode commented:
Converting to draft because it’s late, and I don’t have any more time tonight to dig into the ofborg failure.

wegank commented:
Looks like `pkgsStatic.hello` failed to evaluate on darwin 🤔

wegank commented:
@ofborg build gcc gcc.passthru.tests

reckenrode commented:
> Looks like `pkgsStatic.hello` failed to evaluate on darwin, probably shouldn't `linkManPages llvmPackages.llvm-manpages` there? 🤔

Shouldn’t that have caused an evaluation failure in the bootstrap PR? I’ll look at this more today. There’s also a Rust build failure I was trying to proactively fix that needs fixed too.

reckenrode commented:
Since the fix has been merged into staging-next for the build failure, I’m going to retarget staging since this also touches a separate issue.

reckenrode commented:
@wegank Thanks for pushing the assembler stuff that was added to #242316 but dropped. I’ll add the GCC change to this later today and mark this PR as ready.

wegank commented:
I checked [the failure of `texlive-core-big.bin-2022`](https://malob.github.io/nix-review-tools-reports/nixpkgs:staging-next/nixpkgs_staging-next_1797419.html) this morning and found that `ICU_LIBS` was not correctly set. If the fix for `icu` triggers a world build on darwin, I'm thinking of retargeting this PR on staging-next.

reckenrode commented:
#242519 is the PR to fix the icu issue. I’m waiting on that PR because there are differences between x86_64-darwin and aarch64-darwin in how Libsystem and CF are built, which could cause unwanted paths in the output. I feel pretty confident, but I don’t want to mark it ready when x86_64-darwin could still be broken.

reckenrode committed:
Revert "gcc: fix build on x86_64-darwin"

This reverts commit b39481bc09a6bede802f0bb7b805694ea9cf7f2d.

reckenrode commented:
I pushed a commit to revert the GCC change since this PR addresses the issue generally for clang 11 on x86_64-darwin.

reckenrode commented:
@ofborg build gcc gcc.passthru.tests

reckenrode commented:
@wegank I updated the description since the scope of the PR changed and rebased it on staging-next. The x86_64-darwin stdend still has 460 packages to build on my system before I can confirm there are no issues with the final stdenv.

reckenrode commented:
#242519 is ready for review. It fixes the ICU issue.

vcunat commented:
You have still time for reviewing this; I believe it's possible to include in the current `staging-next` a bit later.

Given our situation, I made Hydra build only `aarch64-darwin` for now, as that will keep it busy by itself for quite some time.  And this PR does no rebuild on `aarch64-darwin`.
